lib/presto-client/index.js

```diff
@@ -220,167 +235,212 @@ Client.prototype.statementResource = function(opts) {
-                };
-
-                /* presto-main/src/main/java/com/facebook/presto/execution/QueryState.java
-                 * QUEUED, PLANNING, STARTING, RUNNING, FINISHED, CANCELED, FAILED
-                 */
-                if (response.stats.state === 'QUEUED'
-                    || response.stats.state === 'PLANNING'
-                    || response.stats.state === 'STARTING'
-                    || response.stats.state === 'RUNNING' && !response.data) {
-                    fetchNextWithTimeout(response.nextUri, client.checkInterval);
-                    return;
-                }
+            first_request = false;
 
-                if (data_callback && response.data) {
-                    data_callback(null, response.data, response.columns, response.stats);
-                }
+            if (state_callback && (last_state !== response.stats.state || enable_verbose_state_callback)) {
+                state_callback(null, response.id, response.stats);
+                last_state = response.stats.state;
+                query_id = response.id;
+            }
 
-                if (response.nextUri) {
-                    fetchNextWithTimeout(response.nextUri, client.checkInterval);
-                    return;
-                }
+            if (columns_callback && response.columns && !columns) {
+                columns = response.columns;
+                columns_callback(null, columns);
+            }
 
-                var finishedStats = response.stats;
+            var fetchNextWithTimeout = function(uri, checkInterval) {
+                setTimeout(function(){ fetch(uri); }, checkInterval);
+            };
 
-                if (fetch_info && response.infoUri) {
-                    client.request(response.infoUri, function(error, code, response){
-                        success_callback(null, finishedStats, response);
-                    });
-                }
-                else {
-                    success_callback(null, finishedStats);
-                }
-            });
-        };
-        fetch_next(firstNextUri);
-    });
+            /* presto-main/src/main/java/com/facebook/presto/execution/QueryState.java
+                * QUEUED, PLANNING, STARTING, RUNNING, FINISHED, CANCELED, FAILED
+                */
+            if (response.stats.state === 'QUEUED'
+                || response.stats.state === 'PLANNING'
+                || response.stats.state === 'STARTING'
+                || response.stats.state === 'RUNNING' && !response.data) {
+                fetchNextWithTimeout(response.nextUri, client.checkInterval);
+                return;
+            }
+
+            if (data_callback && response.data) {
+                data_callback(null, response.data, response.columns, response.stats);
+            }
+
+            if (response.nextUri) {
+                fetchNextWithTimeout(response.nextUri, client.checkInterval);
+                return;
+            }
+
+            clear_timeout();
```

Comment:
Should this be moved up? Right now the docs say that the timeout is counted until the first row of results have come back.

I wonder if it'd be better to have a set timeout until the first results come back, and then have a separate "retry counter" that each 50x error can happen Y times before we abandon the query.